### PR TITLE
tikv: add TsFilter for ResetToVersionWorker

### PR DESCRIPTION
**Pull request by JmPotato**
Signed-off-by: JmPotato <ghzpotato@gmail.com>

### What is changed and how it works?

Issue Number: Ref #13276.

What's Changed:

```commit-message
Add `TsFilter` for `ResetToVersionWorker` to filter out the SST that does not have a newer version than the TS to reset in `CF_WRITE`.
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None.
```

### CODE REVIEW

**Review by longfangsong — APPROVED** [review]

File: src/server/reset_to_version.rs [hunk 1]
```diff
@@ -215,8 +216,16 @@ impl ResetToVersionManager {
         let readopts = IterOptions::new(None, None, false);
         let write_iter = self
             .engine
-            .iterator_opt(CF_WRITE, readopts.clone())
+            .iterator_opt(CF_WRITE, {
+                // Filter out the SST that does not have a newer version than `ts` in
+                // `CF_WRITE`.
+                let mut cf_write_readopts = readopts.clone();
+                cf_write_readopts.set_hint_min_ts(Bound::Excluded(ts.into_inner()));
```
Comment (Connor1996) [inline src/server/reset_to_version.rs:223]: seems should use `hint_max_ts`?

Comment (JmPotato) [reply src/server/reset_to_version.rs:223]: What we need to do here is to filter out the SST whose max TS is smaller than the TS we want to reset, in this case, we don't need to delete any key from it.

Comment (Connor1996) [reply src/server/reset_to_version.rs:223]: oh, my bad

File: src/server/reset_to_version.rs [hunk 2]
```diff
@@ -231,16 +240,23 @@ impl ResetToVersionManager {
...
 
+                let mut total_scanned = 0;
                 while worker.process_next_batch(BATCH_SIZE, &mut wb).expect("reset_to_version failed when removing invalid writes") {
                 }
-                *worker.state.lock()
-                        .expect("failed to lock `ResetToVersionWorker::state` in `ResetToVersionWorker::process_next_batch`")
-                    = ResetToVersionState::RemovingLock { scanned: 0 };
+                {
+                    let mut state = worker.state.lock()
+                            .expect("failed to lock `ResetToVersionWorker::state` in `ResetToVersionWorker::process_next_batch`");
```
Comment (Connor1996) [inline src/server/reset_to_version.rs:248]: `unwrap` is adequate, it's a little verbose here

**Review by Connor1996 — APPROVED** [review]
rest LGTM

File: src/server/reset_to_version.rs [hunk 3]
```diff
@@ -234,33 +234,39 @@ impl ResetToVersionManager {
...
+            thread::Builder::new()
+                .name("reset_to_version".to_string())
+                .spawn_wrapper(move || {
+                    tikv_util::thread_group::set_properties(props);
+                    tikv_alloc::add_thread_memory_accessor();
+
+                    let mut total_scanned = 0;
+                    while worker
+                        .process_next_batch(BATCH_SIZE, &mut wb)
+                        .expect("reset_to_version failed when removing invalid writes")
```
Comment (Connor1996) [inline src/server/reset_to_version.rs:247]: print out the error?

File: src/server/reset_to_version.rs [hunk 4]
```diff
@@ -225,26 +234,39 @@ impl ResetToVersionManager {
...
+                    {
+                    }
+                    {
+                        let mut state = worker.state.lock().unwrap();
+                        total_scanned += *state.scanned();
+                        *state = ResetToVersionState::RemovingLock { scanned: 0 };
```
Comment (Connor1996) [inline src/server/reset_to_version.rs:253]: update the state for each batch of the loop.

File: src/server/reset_to_version.rs [hunk 5]
```diff
@@ -85,10 +83,7 @@ impl ResetToVersionWorker {
 
     fn next_write(&mut self) -> Result<Option<(Vec<u8>, Write)>> {
         if self.write_iter.valid().unwrap() {
```
Comment (Connor1996) [inline src/server/reset_to_version.rs:88]:
```suggestion
        if self.write_iter.valid()? {
```

Comment (Connor1996) [reply src/server/reset_to_version.rs:88]: same for other places

File: src/server/reset_to_version.rs [hunk 6]
```diff
@@ -225,34 +239,43 @@ impl ResetToVersionManager {
...
-            .expect("failed to spawn reset_to_version thread"));
+                    while worker
+                        .process_next_batch(BATCH_SIZE, &mut wb)
+                        .map_err(|err| {
+                            format!(
+                                "reset_to_version failed when removing invalid writes: {}",
+                                err,
+                            )
+                        })
+                        .unwrap()
```
Comment (Connor1996) [inline src/server/reset_to_version.rs:257]: should print the error but avoid panic

**Review by Connor1996 — COMMENTED** [review]
the short value is inlined in write cf, check the commit info first and delete default cf only when needed `box_try!(wb.delete_cf(CF_DEFAULT, default_key.as_encoded()));`

File: src/server/reset_to_version.rs [hunk 7]
```diff
@@ -171,6 +172,11 @@ impl ResetToVersionWorker {
             wb.write().unwrap();
             wb.clear();
         }
+        if !has_more {
```
Comment (Connor1996) [inline src/server/reset_to_version.rs:175]: no need for `has_more` as `wb.is_empty()` indicate it.

File: src/server/reset_to_version.rs [hunk 8]
```diff
@@ -149,15 +152,13 @@ impl ResetToVersionWorker {
...
+                let mut state = self.state.lock().unwrap();
                 debug_assert!(matches!(
                     *state,
                     ResetToVersionState::RemovingLock { scanned: _ }
                 ));
                 *state.scanned() += 1;
+                self.total_scanned += 1;
                 drop(state);
 
                 box_try!(wb.delete_cf(CF_LOCK, self.lock_iter.key()));
```
Comment (Connor1996) [inline src/server/reset_to_version.rs:164]: related default value is not cleaned

File: src/server/reset_to_version.rs [hunk 9]
```diff
@@ -225,34 +231,50 @@ impl ResetToVersionManager {
...
+                            }
+                            Err(e) => {
+                                error!("reset_to_version failed when removing from write/default cfs"; "error" => ?e);
+                                break;
+                            }
+                        }
+                    }
+                    // Delete the lock data in batch then.
+                    loop {
+                        match worker.process_next_batch_lock(&mut wb) {
```
Comment (Connor1996) [inline src/server/reset_to_version.rs:256]: I don't know the purpose of dividing it into two stages. @longfangsong If an error happened in the midst of `process_next_batch`, would it leave an incorrect state?

File: src/server/reset_to_version.rs [hunk 10]
```diff
@@ -225,34 +231,50 @@ impl ResetToVersionManager {
...
+                                    break;
+                                }
+                            }
+                            Err(e) => {
+                                error!("reset_to_version failed when removing from lock cf"; "error" => ?e);
+                                break;
+                            }
+                        }
+                    }
+                    info!("Reset to version done!");
```
Comment (Connor1996) [inline src/server/reset_to_version.rs:268]: print `total_scanned` here

File: src/server/reset_to_version.rs [hunk 11]
```diff
@@ -120,57 +120,55 @@ impl ResetToVersionWorker {
...
+                let mut state = self.state.lock().unwrap();
                 debug_assert!(matches!(
                     *state,
                     ResetToVersionState::RemovingLock { scanned: _ }
                 ));
                 *state.scanned() += 1;
+                self.total_scanned += 1;
                 drop(state);
 
                 box_try!(wb.delete_cf(CF_LOCK, self.lock_iter.key()));
```
Comment (Connor1996) [inline src/server/reset_to_version.rs:156]: User may not pass a `ts` which is larger than the `min-resolved-ts`. What if the primary key is committed, but the secondary lock is left due to error. Better add a paranoid check to make sure the lock's start ts is smaller than provided `ts`.

Comment (Connor1996) [reply src/server/reset_to_version.rs:156]: /cc @MyonKeminta

**Comment by JmPotato** [comment]
Transaction correctness-related comments will be addressed in https://github.com/tikv/tikv/pull/13289.

**Pull request closed without merging**